include/rmm/detail/format.hpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <array>
+#include <cstdio>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+namespace RMM_NAMESPACE {
+namespace detail {
+
+/**
+ * @brief Format a message string with printf-style formatting
+ *
+ * This function performs printf-style formatting to avoid the need for fmt
+ * or spdlog's own templated APIs (which would require exposing spdlog
+ * symbols publicly) and returns the formatted message as a `std::string`.
+ *
+ * @param format The format string
+ * @param args The format arguments
+ */
+template <typename... Args>
+std::string formatted_log(std::string const& format, Args&&... args)
+{
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
+  auto size = static_cast<size_t>(std::snprintf(nullptr, 0, format.c_str(), args...) + 1);
+  if (size <= 0) { throw std::runtime_error("Error during formatting."); }
```

Comment:
Good catch. I copy pasted this from @vyasr and I noticed it may have come from here: https://stackoverflow.com/a/26221725/749748 However that answer just used `int`.  I'm not sure why a cast is needed here at all.

Reply:
> I'm not sure why a cast is needed here at all.

The `size` argument to `snprintf` is a `std::size_t`, so in some scenarios a compiler _might_ warn about the sign conversion (e.g. clang with `-Wsign-conversion`).

Reply:
OK. I tested on compiler explorer with the latest clang and GCC and get no warnings when just using `auto` with no cast.

Reply:
Ha, neither clang nor gcc put `-Wsign-conversion` in `-Wall`.

Reply:
Why would "all" not include all warnings?!

Anyway, your suggestion of ssize_t gives the same warning as int when `Wsign-conversion` is specified, naturally. I also think the original conditional was wrong anyway because it was done after adding 1.

Reply:
> Why would "all" not include all warnings?!

A question for the ages :P.

But yes, mine was also bugged in the same way.